packages/ui/src/components/List/ListContext.tsx

```diff
@@ -196,99 +195,91 @@ export const ListProvider = ({
-          const checkboxIds = Object.keys(checkboxRows)
+            })
 
-          if (state === true) {
-            selectRows(checkboxIds, false)
-          }
-          if ([false, 'indeterminate'].includes(state)) {
-            selectRows(checkboxIds, true)
+            selectRows(checkboxesInRange, currentCheckbox.checked) //  (un)selects the rows in the range
+            setLastCheckedCheckbox(currentCheckbox.value)
           }
-        }
+        } else if (index === 0) setLastCheckedCheckbox(undefined)
 
         /**
          * Handle the case when there is multiple selected value during a time, and the user click without shift event
          */
         setTimeout(() => {
           // clean up
           setInRange(new Set([]))
-          setLastCheckedIndex(checkbox.value)
+          setLastCheckedCheckbox(currentCheckbox.value)
         }, 1)
       }
-
-      const handleOnChange = (checkbox: HTMLInputElement) => {
-        const shouldHandleEvent = inRange.size === 0
-
-        if (shouldHandleEvent) {
-          selectRows([checkbox.value], !checkbox.checked)
-        }
-        setLastCheckedIndex(checkbox.value)
-      }
-
-      refList.current.forEach(checkbox => {
-        function clickHandler(this: HTMLInputElement) {
-          handleClickRange(this)
-        }
-
-        function hoverHandler(this: HTMLInputElement, event: MouseEvent) {
-          handleHover(this, event)
-        }
-
-        function changeHandler(this: HTMLInputElement) {
-          handleOnChange(this)
+      refList.current.forEach((checkbox, index) => {
+        function clickHandler(this: HTMLInputElement, event: MouseEvent) {
+          handleClickRange(this, index, event.shiftKey)
         }
 
-        checkbox.addEventListener('change', changeHandler)
         checkbox.addEventListener('click', clickHandler)
-        checkbox.addEventListener('mouseover', hoverHandler)
 
         handlers.push(() => {
-          checkbox.removeEventListener('change', changeHandler)
           checkbox.removeEventListener('click', clickHandler)
-          checkbox.removeEventListener('mouseover', hoverHandler)
         })
       })
     }
 
     return () => {
       handlers.forEach(cleanup => cleanup())
     }
-  }, [inRange, lastCheckedIndex, selectRows])
+  }, [lastCheckedCheckbox, selectRows])
 
   useEffect(subscribeHandler, [subscribeHandler])
 
+  const handleOnChange = useCallback(
+    (value: string, checked: boolean) => {
+      selectRows([value], !checked)
+      setLastCheckedCheckbox(value)
+    },
+    [selectRows],
+  )
+
+  useEffect(() => {
+    // Re-arrange refList everytime it is modified in order to keep the write order of the elements in it.
+    const allCheckboxes = [
+      ...document.querySelectorAll('input[type="checkbox"]:not([value="all"])'),
+    ].map(element => element as HTMLInputElement)
+    refList.current = allCheckboxes.map(checkbox => checkbox)
+  }, [refList.current.length])
```

Comment:
By doing this you take the risk to select every checkbox in the page. Let's say you have a 2 list in the same page you will select checkboxes for both list while only one is being edited.

We should avoid using a querySelector for interacting with the list / table